src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
+              | Some ((_, prev_addr, _) as prev)
+                when Addr.(equal @@ sibling prev_addr) addr ->
+                  (process_pair height (prev, el) :: processed, None)
+              | Some prev ->
+                  (process_single height prev :: processed, Some el) )
+        in
+        List.rev_append reversed
+        @@ Option.(map ~f:(process_single height) mlast |> to_list)
+      in
+      let rec impl acc height task =
+        let acc' =
+          List.unordered_append (List.map ~f:(fun (a, b, _) -> (b, a)) task) acc
+        in
+        match task with
+        | [] | [ (_, _, []) ] ->
+            acc'
+        | _ ->
+            impl acc' (height + 1) (converge height task)
+      in
+      let hash_account =
+        Option.value_map ~default:Hash.empty_account ~f:Hash.hash_account
+      in
+      Fn.compose (impl [] 0) (List.map ~f:(Tuple3.map_fst ~f:hash_account))
+
+    let finalize_hashes_do t unhashed_accounts =
+      let with_merkle_path_batch accs =
+        let { hashes; _ }, ancestor = maps_and_ancestor t in
+        path_batch_impl
+          ~base_lookup:(Base.merkle_path_batch ancestor)
+          ~self_lookup:
+            (self_merkle_path ~current_location:t.current_location
+               ~depth:t.depth ~hashes )
+          ~fixup_path:(fixup_merkle_path ~hashes)
+          (List.map ~f:snd accs)
+        |> List.map2_exn accs ~f:(fun (a, loc) p ->
+               (a, Location.to_path_exn loc, p) )
+      in
+      let on_snd f (_, a) (_, b) = f a b in
+      List.stable_sort ~compare:(on_snd Location.compare) unhashed_accounts
+      |> List.remove_consecutive_duplicates ~which_to_keep:`First
+           ~equal:(on_snd Location.equal)
+      |> with_merkle_path_batch |> compute_merge_hashes
+      |> List.iter ~f:(Tuple2.uncurry @@ self_set_hash_impl t)
+
+    let finalize_hashes t =
+      let unhashed_accounts = t.unhashed_accounts in
+      if not @@ List.is_empty unhashed_accounts then (
+        t.unhashed_accounts <- [] ;
+        finalize_hashes_do t unhashed_accounts )
+
+    (** Either copies accumulated or initializes it with the parent being used as the [base]. *)
+    let to_accumulated t =
+      finalize_hashes t ;
+      actualize_accumulated t ;
+      match (t.accumulated, t.parent) with
+      | Some { base; detached_next_signal; next; current }, _ ->
+          { base; detached_next_signal; next; current }
+      | None, Ok base ->
+          { base
+          ; next = t.maps
+          ; current = t.maps
+          ; detached_next_signal = t.detached_parent_signal
+          }
+      | None, Error loc ->
+          raise (Dangling_parent_reference (t.uuid, loc))
+
+    let self_set_hash t address hash =
+      finalize_hashes t ;
+      self_set_hash_impl t address hash
+
     let set_inner_hash_at_addr_exn t address hash =
       assert_is_attached t ;
       assert (Addr.depth address <= t.depth) ;
       self_set_hash t address hash
 
+    let hashes_and_ancestor t =
```

Comment:
Note to myself: mostly unchanged except for the first line. `maps_and_ancestor` was left as it is, and `hashes_and_ancestor` additionally runs `finalize_hashes` before the sub-call.